src/coreclr/src/binder/utils.cpp

```diff
@@ -115,4 +117,139 @@ namespace BINDER_SPACE
+            return S_FALSE;
+        }
+
+        // Support paths being wrapped with quotations
+        if (paths.Skip(startPos, W('\"')))
+        {
+            wrappedWithQuotes = true;
+        }
+
+        SString::Iterator iEnd = startPos;      // Where current path ends
+        SString::Iterator iNext;                // Where next path starts
+        if (wrappedWithQuotes)
+        {
+            if (paths.Find(iEnd, W('\"')))
+            {
+                iNext = iEnd;
+                // Find where the next path starts - there should be a path separator right after the closing quotation mark
+                if (paths.Find(iNext, PATH_SEPARATOR_CHAR_W))
+                {
+                    iNext++;
+                }
+                else
+                {
+                    iNext = paths.End();
+                }
+            }
+            else
+            {
+                // There was no terminating quotation mark - that's bad
+                GO_WITH_HRESULT(E_INVALIDARG);
+            }
+        }
+        else if (paths.Find(iEnd, PATH_SEPARATOR_CHAR_W))
+        {
+            iNext = iEnd + 1;
+        }
+        else
+        {
+            iNext = iEnd = paths.End();
+        }
+
+        // Skip any trailing spaces
+        while (iEnd[-1] == W(' '))
+        {
+            iEnd--;
+        }
+
+        _ASSERTE(startPos < iEnd);
+
+        outPath.Set(paths, startPos, iEnd);
+        startPos = iNext;
+    Exit:
+        return hr;
+    }
+
+    HRESULT GetNextTPAPath(SString& paths, SString::Iterator& startPos, bool dllOnly, SString& outPath, SString& simpleName, bool& isNativeImage)
+    {
+        HRESULT hr = S_OK;
+        isNativeImage = false;
+
+        HRESULT pathResult = S_OK;
+        IF_FAIL_GO(pathResult = GetNextPath(paths, startPos, outPath));
+        if (pathResult == S_FALSE)
+        {
+            return S_FALSE;
+        }
+
+#ifndef CROSSGEN_COMPILE
+        if (Path::IsRelative(outPath))
+        {
+            GO_WITH_HRESULT(E_INVALIDARG);
+        }
+#endif
+
+        {
+            // Find the beginning of the simple name
+            SString::Iterator iSimpleNameStart = outPath.End();
+
+            if (!outPath.FindBack(iSimpleNameStart, DIRECTORY_SEPARATOR_CHAR_W))
+            {
+                iSimpleNameStart = outPath.Begin();
+            }
+            else
+            {
+                // Advance past the directory separator to the first character of the file name
+                iSimpleNameStart++;
+            }
+
+            if (iSimpleNameStart == outPath.End())
+            {
+                GO_WITH_HRESULT(E_INVALIDARG);
+            }
+
+            // GCC complains if we create SStrings inline as part of a function call
+            SString sNiDll(W(".ni.dll"));
+            SString sNiExe(W(".ni.exe"));
+            SString sDll(W(".dll"));
+            SString sExe(W(".exe"));
+
+            if (!dllOnly && (outPath.EndsWithCaseInsensitive(sNiDll) ||
```

Comment:
Why do we exclude the .ni.dll here? I wonder if that's correct.

Reply:
The `dllOnly` is currently only set to `true` when searching for System.Private.CoreLib (in the added fallback logic), so I believe the handling of it was more targeted to avoiding extra work in that scenario of just looking for .dll and not .ni.dll. For longer term / 6.0, I expect it will be refactored away per https://github.com/dotnet/runtime/pull/42435#discussion_r493857718.